### PR TITLE
Multiple categories

### PR DESCRIPTION
**Pull request by adishalev**
UCCA now supports: 
1. annotations with more than one category per edge.
2. annotations from refinement tasks

* back-supporting old xml format

### CODE REVIEW

File: ucca/convert.py [hunk 1]
```diff
@@ -821,6 +836,18 @@ def get_json_attrib(d):
...
 
 
+def get_categories_details(d):
+    # mapping from a category id to the category's parent, layer
+    curr_layer = d['project']['layer']
+    categories = {}
+    while curr_layer:
+        base_layer = curr_layer['name']
+        for c in curr_layer['categories']:
+            categories[c['id']] = {'name': c["name"], 'parent': c['parent'] if 'parent' in c else None, 'layer': base_layer}
```
Comment (danielhers) [inline ucca/convert.py]: `c['parent'] if 'parent' in c else None` can be replaced by `c.get("parent")`

File: ucca/convert.py [hunk 2]
```diff
@@ -879,51 +907,54 @@ def from_json(lines, *args, skip_category_mapping=False, by_external_id=False, *
...
-                category_name = category.get("name") or category_id_to_name[category["id"]]
+                category_name = category.get("name") or categories[category["id"]]['name']
             except KeyError as e:
                 raise ValueError("Category missing from layer: " + category["id"]) from e
-            tag = category_name_to_edge_tag.get(category_name.replace(" ", ""), category_name)
-            tags.append(tag)
-
-        if not tags:
+            c_tag = category_name_to_edge_tag.get(category_name.replace(" ", ""), category_name.replace(" ", "_"))
+            c_slot = category["slot"] if 'slot' in category else ""
```
Comment (danielhers) [inline ucca/convert.py]: `category["slot"] if 'slot' in category else ""` can be replaced by `category.get("slot", "")`

File: ucca/convert.py [hunk 3]
```diff
@@ -879,51 +907,54 @@ def from_json(lines, *args, skip_category_mapping=False, by_external_id=False, *
...
+            c_layer = categories[category["id"]]['layer']
```
Comment (danielhers) [inline ucca/convert.py]: Better extract `categories[category["id"]]` to a variable

File: ucca/convert.py [hunk 4]
```diff
@@ -879,51 +907,54 @@ def from_json(lines, *args, skip_category_mapping=False, by_external_id=False, *
...
+            if c_layer == base_layer:
+                base_slot = c_slot
+            c_parent = categories[category["id"]]['parent']
+            if c_parent:   # make sure it is not empty
+                c_parent = c_parent['name']
```
Comment (danielhers) [inline ucca/convert.py]: Shouldn't the parent category name be mapped to an edge tag too?

File: ucca/convert.py [hunk 5]
```diff
@@ -879,51 +907,54 @@ def from_json(lines, *args, skip_category_mapping=False, by_external_id=False, *
...
+            unit_categories.append((c_tag, c_slot, c_layer, c_parent))
```
Comment (danielhers) [inline ucca/convert.py:925]: Why not create `Category` objects rather than dealing with 4-tuples everywhere?

File: ucca/core.py [hunk 6]
```diff
@@ -187,13 +187,52 @@ def items(self):
...
+
+
 class Edge:
     """Labeled edge between two :class:`Node` objects in UCCA annotation graph.
 
     An edge between Nodes in a :class:`Passage` is a simple object; it is a
     directed edge whose ID is derived by the parent and child of the edge,
     it is mostly immutable except for its attributes, and it is labeled with
-    the connection type between the Nodes.
+    the connection type between the Nodes. an edge can have multiple annotations, representing
```
Comment (danielhers) [inline ucca/core.py]: Capitalization

File: ucca/core.py [hunk 7]
```diff
@@ -248,6 +288,15 @@ def root(self):
     def parent(self):
         return self._parent
 
+    @property
+    def categories(self):
+        return self._categories
+
+    @categories.setter
+    def categories(self, new_categories):
```
Comment (danielhers) [inline ucca/core.py:300]: Since it's a setter, I would expect it to *replace* the list of categories rather than adding to it

File: ucca/core.py [hunk 8]
```diff
@@ -286,9 +335,32 @@ def equals(self, other, *, recursive=True, ordered=False,
...
+        """Iterates the :class:`Edge` objects in the subtree of self.
+
+        :param obj: yield Category objects
+
+        Yields:
+            a :class:`Node` or :class:`Edge` object according to the iteration
+            parameters.
+
+        """
+        for category in self._categories:
```
Comment (danielhers) [inline ucca/core.py]: Can be replaced by `yield from self._categories`.

File: ucca/core.py [hunk 9]
```diff
@@ -286,9 +335,32 @@ def equals(self, other, *, recursive=True, ordered=False,
...
+        return c
+
     def __repr__(self):
         return self.ID
 
+    def __getitem__(self, index):
+        return self._categories[index]
+
+    def iter(self):
+        """Iterates the :class:`Edge` objects in the subtree of self.
```
Comment (danielhers) [inline ucca/core.py]: "subtree" isn't correct here

File: ucca/core.py [hunk 10]
```diff
@@ -286,9 +335,32 @@ def equals(self, other, *, recursive=True, ordered=False,
...
+
+        :param obj: yield Category objects
```
Comment (danielhers) [inline ucca/core.py]: There is no such param in this method

File: ucca/core.py [hunk 11]
```diff
@@ -286,9 +335,32 @@ def equals(self, other, *, recursive=True, ordered=False,
...
+
+        Yields:
+            a :class:`Node` or :class:`Edge` object according to the iteration
```
Comment (danielhers) [inline ucca/core.py]: Should be Category objects

File: ucca/core.py [hunk 12]
```diff
@@ -400,10 +472,10 @@ def __repr__(self):
         return Node.__name__ + "(" + self.ID + ")"
 
     @ModifyPassage
-    def add(self, edge_tag, node, *, edge_attrib=None):
+    def add(self, edge_categories, node, *, edge_attrib=None):
         """Adds another :class:`Node` object as a child of self.
 
-        :param edge_tag: the label of the :class:`Edge` connecting between the
+        :param edge_categories: the labels of the :class:`Edge` connecting between the
```
Comment (danielhers) [inline ucca/core.py]: Need more informative documentation on the expected type of the list elements (4-tuples) if we keep it this way

File: ucca/core.py [hunk 13]
```diff
@@ -211,7 +251,7 @@ class Edge:
 
     ID_FORMAT = "{}->{}"
 
-    def __init__(self, root, tag, parent, child, attrib=None):
```
Comment (danielhers) [inline ucca/core.py:214]: How about optionally allowing the `tag` parameter as a shortcut for adding a single category?

File: ucca/core.py [hunk 14]
```diff
@@ -829,6 +909,14 @@ def layers(self):
     def nodes(self):
         return self._nodes.copy()
 
+    @property
+    def categories(self):
+        return self._categories
```
Comment (danielhers) [inline ucca/core.py]: I think this and `refined_categories` should copy the attributes (by calling `dict()` and `list()` respectively), like the other getters

File: ucca/core.py [hunk 15]
```diff
@@ -938,6 +1026,14 @@ def _add_layer(self, layer):
             raise DuplicateIdError(layer.ID)
         self._layers[layer.ID] = layer
 
+    @ModifyPassage
+    def _update_categories(self, category):
+        self._categories[category.tag] = {"layer": category.layer, "slot": category.slot, "parent": category.parent}
```
Comment (danielhers) [inline ucca/core.py:1040]: Why a dict and not Category object?

File: ucca/layer1.py [hunk 16]
```diff
@@ -433,12 +433,12 @@ def next_id(self):
...
                 return id_str
 
-    def add_fnode(self, parent, edge_tag, *, implicit=False):
+    def add_fnode(self, parent, edge_categories, *, implicit=False):
         """Adds a new :class:`FNode` whose parent and Edge tag are given.
 
         :param parent: the FNode which will be the parent of the new FNode.
                 If the parent is None, adds under the layer head FNode.
-        :param edge_tag: the tag on the Edge between the parent and the new FNode.
+        :param edge_categories: list of categories on the Edge between the parent and the new FNode.
```
Comment (danielhers) [inline ucca/layer1.py:441]: Again, document the exact expected format of the parameter (tuples or Category objects)

File: ucca/convert.py [hunk 17]
```diff
@@ -879,51 +907,54 @@ def from_json(lines, *args, skip_category_mapping=False, by_external_id=False, *
...
+                node.extra['remarks'] = comment
+            for token in children_tokens:
+                token_id_to_preterminal[token["id"]] = node
 
     # Attach terminals to non-terminals
     for token_id, node in token_id_to_preterminal.items():
         terminal = token_id_to_terminal[token_id]
         if skip_category_mapping or not layer0.is_punct(terminal):
-            node.add(EdgeTags.Terminal, terminal)
+            node.add([(EdgeTags.Terminal, base_slot, base_layer, "")], terminal)
```
Comment (danielhers) [inline ucca/convert.py]: `""` for parent can be omitted since it's optional (same for other places in the code)

File: ucca/normalization.py [hunk 18]
```diff
@@ -108,7 +108,7 @@ def move_sub_scene_elements(node):
 def separate_scenes(node, l1, top_level=False):
     if (node.is_scene() or node.participants) and (top_level or node.parallel_scenes):
         edges = list(node)
-        scene = l1.add_fnode(node, ETags.ParallelScene)
+        scene = l1.add_fnode(node, [(ETags.ParallelScene)])
```
Comment (danielhers) [inline ucca/normalization.py]: Redundant parentheses - if you want a 1-tuple, you must add a comma: `(ETags.ParallelScene,)`. Same for other places in the code

File: uccaapp/export_units_by_filter.py [hunk 19]
```diff
@@ -21,25 +22,29 @@ def get_top_level_ancestor(node):
...
+def main(output=None, comment=False, categories=(), **kwargs):
     filtered_nodes = []
     for passage, task_id, user_id in TaskDownloader(**kwargs).download_tasks(**kwargs, write=False):
         for node in passage.layer(layer1.LAYER_ID).all:
             if comment and node.extra.get("remarks"):
-                filtered_nodes.append(("comment",node,task_id,user_id))
+                filtered_nodes.append(("comment", node, task_id, user_id))
             else:
-                all_tags = node.extra.get("all_tags")
+                all_tags = []
```
Comment (danielhers) [inline uccaapp/export_units_by_filter.py:71]: This is fine, but FYI, you could write it as `all_tags = [c.tag for edge in node for c in edge.categories]`

**Review by danielhers — CHANGES_REQUESTED** [review]
Thanks, this is an important change. Please see my comments. Also, I'd like to have the tests pass before merging: https://travis-ci.org/danielhers/ucca/jobs/479585327 (you can run them offline with pytest: `pytest ucca/tests`).

**Comment by danielhers** [comment]
@adishalev let me know if you need help with the fixes (to the tests etc.).

**Comment by danielhers** [comment]
Looks good, I think. Just need to update `test_files/standard3.xml` to be in the new format.

File: ucca/core.py [hunk 20]
```diff
@@ -222,22 +266,26 @@ def __init__(self, root, tag, parent, child, attrib=None):
...
         self.extra = {}
 
     @property
     def tag(self):
-        return self._tag
+        return self._categories[0].tag
+
+    #@tag.setter
+    #def tag(self, new_tag):
+    #    self._categories[0].tag = new_tag
```
Comment (danielhers) [inline ucca/core.py]: Is this commented-out on purpose?

Comment (adishalev) [reply ucca/core.py]: you can remove it

**Comment by adishalev** [comment]
I changed the code to support adding an edge with a single tag by moving the code that handles multiple categories to core.Node.add_multiple() and layer1.Fnode.add_fnode_multiple(). it fixed most of the test failures.

All the tests pass now, except for test_to_standard() which fails because the xml that was loaded from site has \&quot; while the loaded ref has "" in the extra remarks.
<extra remarks="\&quot;remark\&quot;"
<extra remarks="remark"
I'm not sure which one is correct, could you please take a look at this?

**Comment by adishalev** [comment]
It's ok to leave test_files/standard3.xml as is, it is converted to the new format as part of the test.

**Comment by danielhers** [comment]
Looks like `from_standard` strips the quotes. It didn't fail before because it just wasn't called in the test.
It strips the quotes because it tries to read the `extra` entries as JSON, and if it succeeds, it returns the decoded JSON. Otherwise it returns the original string. Just `remarks` is not valid JSON, but `"remarks"` is, and is decoded to the string `remarks`... I'd like to keep this JSON decoding capability (since sometimes we save actual dicts of values to `extra` entries), so I think I'll just remove the extra `&quot;`s from `standard3.xml`.

**Comment by danielhers** [comment]
Thanks @adishalev ! I respect your choice to keep values "primitive" and create `Category` objects only internally, but I wonder what the considerations were.
By the way, I think you should configure your email with `git config --global user.email`, because right now the commits are not associated with your GitHub username.

**Comment by adishalev** [comment]
I left the comments about primitives vs. categories open as I plan to change those primitives to categories as part of my next commits. 
Thanks for your tip about the git email configuration!

**Comment by danielhers** [comment]
I see. It seemed pretty good already though so I merged it, but I'll be glad to merge another pull request when you're ready.

**Pull request merged**